### ISSUE TITLE
`_base_est_fit_predict` looks wrong

### ISSUE THREAD

**Issue by jonathanng** [civisanalytics/civisml-extensions#16]
I believe cv.split(X, y) might give different results when called at different times. There might be randomness involved like shuffling, etc.

This is problematic because `ybase`, which is based on the old `train` & `test` but not coincide with with the new `test` and therefore `y[test]`.

old `train` & `test`:

```
for train, test in cv.split(X, y):
    for name, est in self.estimator_list[:-1]:
        # adapted from sklearn.model_selection._fit_and_predict
        # Adjust length of sample weights
        fit_params_est_adjusted = dict([
            (k, _index_param_value(X, v, train))
            for k, v in fit_params_ests[name].items()])

        # Fit estimator on training set and score out-of-sample
        _jobs.append(delayed(_fit_predict)(
            clone(est),
            X[train],
            y[train],
            X[test],
            **fit_params_est_adjusted))
```

new `train` & `test`:

```
# Extract the results from joblib
Xmeta, ymeta = None, None
for train, test in cv.split(X, y):
    ybase = np.empty((y[test].shape[0], 0))
    for name, est in self.estimator_list[:-1]:
        # Build design matrix out of out-of-sample predictions
        ybase = np.hstack((ybase, _out.pop(0)))

    # Append the test outputs to what will eventually be the features
    # for the meta-estimator.
    if Xmeta is not None:
        ymeta = np.concatenate((ymeta, y[test]))
        Xmeta = np.vstack((Xmeta, ybase))
    else:
        Xmeta = ybase
        ymeta = y[test]
```

**Comment by beckermr** [comment]
Good catch! I think you are correct. 

However, this looks to be true only if `shuffle=True`. See the code [here](https://github.com/scikit-learn/scikit-learn/blob/a24c8b46/sklearn/model_selection/_split.py#L425). 

Here is a quick example:

With `shuffle=False` we get:

```python
In [1]: from sklearn.model_selection import check_cv

In [2]: cv = check_cv(3)

In [3]: cv
Out[3]: KFold(n_splits=3, random_state=None, shuffle=False)

In [4]: for trn, tst in cv.split(np.arange(9).reshape(-1, 1)):
   ...:     print("trn:", trn)
   ...:     print("tst:", tst)
   ...:     
trn: [3 4 5 6 7 8]
tst: [0 1 2]
trn: [0 1 2 6 7 8]
tst: [3 4 5]
trn: [0 1 2 3 4 5]
tst: [6 7 8]

In [5]: for trn, tst in cv.split(np.arange(9).reshape(-1, 1)):
   ...:     print("trn:", trn)
   ...:     print("tst:", tst)
   ...:     
trn: [3 4 5 6 7 8]
tst: [0 1 2]
trn: [0 1 2 6 7 8]
tst: [3 4 5]
trn: [0 1 2 3 4 5]
tst: [6 7 8]
``` 

With `shuffle=True` we get:

```python
In [8]: cv = KFold(n_splits=3, shuffle=True)

In [9]: for trn, tst in cv.split(np.arange(9).reshape(-1, 1)):
   ...:     print("trn:", trn)
   ...:     print("tst:", tst)
   ...:     
trn: [1 4 5 6 7 8]
tst: [0 2 3]
trn: [0 1 2 3 6 7]
tst: [4 5 8]
trn: [0 2 3 4 5 8]
tst: [1 6 7]

In [10]: for trn, tst in cv.split(np.arange(9).reshape(-1, 1)):
    ...:     print("trn:", trn)
    ...:     print("tst:", tst)
    ...:     
trn: [1 2 3 6 7 8]
tst: [0 4 5]
trn: [0 2 3 4 5 8]
tst: [1 6 7]
trn: [0 1 4 5 6 7]
tst: [2 3 8]
```

I think this explains why the tests currently pass, etc.

**Comment by kcrum** [comment]
Thanks for reporting this @jonathanng. There's a bug fix on the way, but for now if you pass `shuffle=True` to your cross-validator, please explicitly seed the random number generator by passing something to the `random_state` kwarg. I believe the `cv.split` call will yield the same indices twice when `shuffle=True` AND `random_state=<something NOT equal to None>`.
